refactor: stringify

RunDevelopment opened the pull request:
I refactored the stringify code. It now clearly distinguishes the 3 formatting cases (description, example, regular tag).

Changes:
- `formatDescription` no longer decides which texts are formatted. Stringify now decides that.
- Stringify now creates the whole comment content (= comment value) instead of being called tag by tag.
- `SPACE_TAG_DATA` has been removed. Spacing between tags is now handled via the `insertLine{Before,After}` properties of a `TagStringifyResult`.
- A single-line JSDoc comment (e.g. `/** @type {number} */`) will now only be created if the comment length does not exceed `printWidth`.
- Alignment offsets are now calculated per group. This means that the properties of `@typedef`s can be aligned differently. I personally think this is better but I will revert it if wanted. Just tell me @hosseinmd.

There are also a few other changes that I will explain via comments in the test files.

RunDevelopment commented on tests/__snapshots__/files/prism-core.js.shot:
```diff
@@ -278,11 +278,11 @@ var Prism = (function (_self) {
-			 *     // CSS doesn't have a 'color' token, so this token will be appended
-			 *     'color': /\\\\b(?:red|green|blue)\\\\b/
+			 *       // Prism.languages.css already has a 'comment' token, so this token will overwrite CSS' 'comment' token
+			 *       // at its original position
```
The indentation of example code is now properly detected and removed. In Prism's case, the example code wasn't indented with 2 spaces before, so nothing is removed.

RunDevelopment commented on tests/__snapshots__/files/prism-core.js.shot:
```diff
@@ -464,10 +464,12 @@ var Prism = (function (_self) {
 		 * @memberof Prism
-		 * @param {ParentNode} container The root element, whose descendants that have a \`.language-xxxx\` class will be highlighted.
+		 * @param {ParentNode} container
+		 *   The root element, whose descendants that have a \`.language-xxxx\` class will be highlighted.
```
Cases similar to this one can be seen in a few other places as well. They are caused by [this function allowing for 10 extra chars](https://github.com/hosseinmd/prettier-plugin-jsdoc/blob/54bdbc54614d2919f426a926b0612de47d5e5fdd/src/descriptionFormatter.ts#L219) in the last line so the description doesn't get broken into 2 lines as it should be. 

`stringify` checks and sees that the tag will go over `printWidth` if it appends the description to the rest of the tag as is. Since the description is too long to fit, `stringify` will add a line break before the description. This is why this is broken in such a weird way.

The fix is simple. Don't allow the extra 10 chars.

RunDevelopment commented on tests/__snapshots__/singleTag.test.js.snap:
```diff
@@ -5,7 +5,9 @@ exports[`single tag 1`] = `
 export const SubDomain = {
-  /** @returns {import(\\"axios\\").AxiosResponse<import(\\"../types\\").SubDomain>} */
+  /**
+   * @returns {import(\\"axios\\").AxiosResponse<import(\\"../types\\").SubDomain>}
```
As I said in the description of this PR:

> A single-line JSDoc comment (e.g. `/** @type {number} */`) will now only be created if the comment length does not exceed `printWidth`.

hosseinmd replied:
This should be a single line, we created a single line if there is a single line in comment that is not about printwidth, we don't strict about printwidth, that line could be more than printwidth, that is not matter.

RunDevelopment replied:
The user specifically set the print width, so we should honor that if possible IMO. Otherwise, why have a print width in the first place?

RunDevelopment replied:
Or is there a reason why we should ignore print width?

RunDevelopment replied:
I removed the change.

RunDevelopment commented on tests/__snapshots__/files/prism-core.js.shot:
```diff
@@ -587,8 +589,8 @@ var Prism = (function (_self) {
 		/**
-		 * Low-level function, only use if you know what you’re doing. It accepts a string of text as input and the
-		 * language definitions to use, and returns a string with the HTML produced.
+		 * Low-level function, only use if you know what you’re doing. It accepts a string of text as input and the language
```
I don't know why this changed. It seems like something got fixed along the way but I can't tell you what. However, the change is correct. The previous text was wrapped too early.

hosseinmd commented:
I like simple code, keep code simple and focus on fix a problem, your PR just added complexity to code.

RunDevelopment commented:
I moved some code and made some simplifications. Please take a look at it again.

> I like simple code

I do too. That's why I rewrote `stringify`. Right now, this is what a plugin I have installed says about `stringify`:

![image](https://user-images.githubusercontent.com/20878432/107078183-3be9fd80-67ee-11eb-93ca-6167eb2bdb57.png)

RunDevelopment commented:
@hosseinmd Could you please re-review this?

hosseinmd commented:
Test doesn't pass. Please fix them.

RunDevelopment commented:
What? If they didn't pass, I wouldn't have been able to make the commit because of husky. What tests don't pass?

It may be time for a Github action CI...

hosseinmd commented:
You changed snapshot to wrong result

RunDevelopment commented:
> You changed snapshot to wrong result

As I said, the tests pass on my machine. I don't know which one specifically doesn't pass. Could you please show me the console output of the failed test?

RunDevelopment commented:
Ok, I just setup an new Ubuntu system and tried `npm test` on that one. Tests pass. Also, I never manually edit snapshots, I always use the `npm test -- -u` command.

I am now fairly certain that either we have a misunderstanding or that there is something wrong with your setup.

hosseinmd commented:
Never use `-u`. That is for replace new snapshot, use it when you know the old snapshot is wrong

RunDevelopment commented:
Ok, now I get it, I misunderstood. When you said "Test doesn't pass" (= `npm test` fails), you really meant that the snapshot files shouldn't have been changed by this PR. I should have caught that after your second message. I'm sorry for being too slow there.

With that out of the way: The shapshots changed in 4 different ways. 
1. [Example indentation.](https://github.com/hosseinmd/prettier-plugin-jsdoc/pull/59#discussion_r571061160) (intended)
2. [Single-line comments greater than print width.](https://github.com/hosseinmd/prettier-plugin-jsdoc/pull/59#discussion_r571067039) (intended)
3. [Word breaking.](https://github.com/hosseinmd/prettier-plugin-jsdoc/pull/59#discussion_r571069849) (fixing previous error)
4. [Single-line descriptions greater than print width.](https://github.com/hosseinmd/prettier-plugin-jsdoc/pull/59#discussion_r571065914) (unintended; for the full explanation, see the link)

You seem to not like 2), so I will take it out of this PR and make a new one for it after this one.

hosseinmd commented on src/parser.ts:
```diff
@@ -440,3 +372,31 @@ function getParamsOrders(ast: AST, tokenIndex: number) {
+ */
+function combineName(tag: commentParser.Tag): void {
+  // Additional operations on name
+  if (tag.name) {
```
Please destruct tag like `function combineName({tag,name,optional,default:_default}: commentParser.Tag)`

RunDevelopment replied:
The function changes the tag object itself. Destructuring is not possible.

hosseinmd commented on src/parser.ts:
```diff
@@ -149,91 +149,23 @@ export const getParser = (parser: Parser["parse"]) =>
-      comment.value = comment.value.trimEnd();
-      if (comment.value) {
-        comment.value = addStarsToTheBeginningOfTheLines(comment.value);
+      const content = stringifyCommentContent(groups, {
```
1- I think that is not a good idea to move mapping to stringify don't do that.
2- stringify renamed bad, we are working on comment no need to mention it when naming variables or functions

RunDevelopment replied:
> move mapping to stringify

What do you mean by that? What mapping?

> stringify renamed

I did that because it's not clear what conceptual unit is being stringified. What does `stringify` do? Does it stringify a single tag, a group of tags, the whole comment, or only the content of the comment? `stringify` alone doesn't say much, so I specified what was being stringified.

hosseinmd reviewed (commented):
Please don't feel hard, I will close this PR, there is a lot of things I can't mention all please start a new PR and give me small changes to be discussable.

RunDevelopment commented:
Sure thing! I'll split it into many smaller ones again.
